### DISCUSSION TITLE
remark-language-server: autosuggestions

### DISCUSSION THREAD

**Discussion by Murderlon** [remarkjs/remark#940] [Ideas]
[Latest code insiders has autosuggestions for markdown](https://twitter.com/mattbierner/status/1483590058066149376).

Seems to have:
- Path / file links / image links
- Header links 
- Reference links

Wondering if this is possible and wanted in remark-language-server. AFAIK this is a bit different, as normally the language server just takes your remark config, it doesn't determine in itself what to lint/do. But I can imagine this would require (opinionated) plugins to be baked into the language server to understand enough to give suggestions, and crawling the file system.

Does this make sense to implement?

cc @remcohaszing

**Comment by remcohaszing** [comment]
This functionality isn’t related to the `unified-language-server` functionality to provide formatting and linting based on `unified-engine`. I have already been thinking about the scenario that `unified-language-server` can accept a function to enhance the default functionality, i.e. provide markdown specific code completions.

These functionalities are likely unrelated to `unified-engine`. I.e. it’s still useful to provide completions based in ignored files.

```js
import {createUnifiedLanguageServer} from 'unified-language-server'

createUnifiedLanguageServer({
  ignoreName: '.remarkignore',
  packageField: 'remarkConfig',
  pluginPrefix: 'remark',
  processorName: 'remark',
  processorSpecifier: 'remark',
  rcName: '.remarkrc',
  enhance(connection, textDocuments, processor) {
    connection.on/* some event */((event) => {
      // handle it
    })
  }
})
```

I’m not sure about this specific functionality though. VS Code already provides it. This means the functionality would be duplicated for VS Code users, but I do see why this is useful for other editors.

**Reply by Murderlon** [reply]
> I’m not sure about this specific functionality though. VS Code already provides it. This means the functionality would be duplicated for VS Code users, but I do see why this is useful for other editors.

I definitely think it's still valuable. It seems to be locked in VS Code in what it does but more importantly it's not in other editors.

> I have already been thinking about the scenario that unified-language-server can accept a function to enhance the default functionality, i.e. provide markdown specific code completions.

Seems interesting. Although probably intense to write yourself if you want something like this, so I can imagine that to be 'plugins' too?

> These functionalities are likely unrelated to `unified-engine`. I.e. it’s still useful to provide completions based in ignored files.

Good point, but for a project you probably set up remark for all the markdown files that matter. So just having autocomplete there instead of in every markdown file seems okay in my opinion.

**Reply by wooorm** [reply]
One part of the conversation is whether this feature makes sense:
a) I agree that it’s needed for other editors,
b) a (remark) AST based system can be smarter and do more than links,
c) when adding a markdown previewer, we are already improving existing features

The other part is, can it work, and how? Here’s one idea:
- [completion requests](https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#textDocument_completion), so we have the caret’s position
- we can combine that position with the AST to find the list of smallest and biggest nodes that contain it (e.g., `[link, paragraph, root]`)
- because we’re in a link, autocomplete from the file system (note: what other things can we do on other conditions?)

There’s a problem with ASTs though:
a) it only works for valid constructs (`[a](|)` is in a link, `[a](|` is not in a link),
b) it may not be detailed enough (for a cursor in `[a](b)`, it gets a bit hacky where the caret is)

An alternative idea is to use the tokens from micromark, that way we can know a lot more?

> > These functionalities are likely unrelated to unified-engine. I.e. it’s still useful to provide completions based in ignored files.
>
> Good point, but for a project you probably set up remark for all the markdown files that matter. So just having autocomplete there instead of in every markdown file seems okay in my opinion.

I think this is similar to why I feel the markdown previewer should also be “agnostic” to `.remarkrc`s and such. Maybe both the previewer and this completion functionality can work on any markdown file and ignore whether and how the project is configured?

**Reply by ChristianMurphy** [reply]
>  it only works for valid constructs `([a](|)` is in a link, `[a](|` is not in a link),

This could be eased with auto syntax (bracket) closing.
VSCode supports this natively https://code.visualstudio.com/api/language-extensions/language-configuration-guide#autoclosing
LSP will hopefully support this soon https://github.com/microsoft/language-server-protocol/issues/672

**Reply by Murderlon** [reply]
> This could be eased with auto syntax (bracket) closing.

I wouldn't bet this implementation on it. I, and many others, don't use auto closing.
